acp/acp_local.go

```diff
@@ -0,0 +1,328 @@
+	"github.com/valyala/fastjson"
+
+	"github.com/sourcenetwork/defradb/errors"
+)
+
+var (
+	_ ACPModule = (*ACPLocal)(nil)
+)
+
+// ACPLocal represents a local acp module implementation that makes no remote calls.
+type ACPLocal struct {
+	pathToStore string
+	localModule *embedded.LocalACP
+}
+
+// NewLocalACPModule returns an Option with the underlying acp module started (ACPModule.Start() called)
+// if successfully initialized, otherwise the `Option.HasValue() == false`.
+//
+// The caller is responsible to free the resources by calling the `ACPModule.Close()` on the underlying
+// acp module if `Option.HasValue() == true`, i.e. the acp module was initialized.
+func NewLocalACPModule(ctx context.Context, rootDir string) immutable.Option[ACPModule] {
+	var acpModule ACPLocal
+
+	if err := acpModule.Start(ctx, rootDir); err != nil {
+		return NoACPModule
+	}
+
+	return immutable.Some[ACPModule](&acpModule)
+}
+
+func (l *ACPLocal) Start(ctx context.Context, rootDir string) error {
+	acpStorePath := rootDir + "/" + embedded.DefaultDataDir
+
+	var localACPModule embedded.LocalACP
+	var err error
+
+	if rootDir == "" { // Use a non-persistent, i.e. in memory store.
+		localACPModule, err = embedded.NewLocalACP(
+			embedded.WithInMemStore(),
+		)
+	} else {
+		localACPModule, err = embedded.NewLocalACP(
+			embedded.WithPersistentStorage(acpStorePath),
+		)
+	}
+
+	if err != nil {
+		return NewErrInitializationOfACPFailed(err, "Local", rootDir)
+	}
+
+	l.localModule = &localACPModule
+	l.pathToStore = acpStorePath
+
+	return nil
+}
+
+func (l *ACPLocal) Close() error {
+	return l.localModule.Close()
+}
+
+func (l *ACPLocal) AddPolicy(
+	ctx context.Context,
+	creatorID string,
+	policy string,
+) (string, error) {
+	// Having a creator identity is a MUST requirement for adding a policy.
+	if creatorID == "" {
+		return "", ErrPolicyCreatorMustNotBeEmpty
+	}
+
+	if policy == "" {
+		return "", ErrPolicyDataMustNotBeEmpty
+	}
+
+	var createPolicy types.MsgCreatePolicy
+
+	// Assume policy is in YAML format by default.
+	policyMarshalType := types.PolicyMarshalingType_SHORT_YAML
+	if isJSON := fastjson.Validate(policy) == nil; isJSON { // Detect JSON format.
+		policyMarshalType = types.PolicyMarshalingType_SHORT_JSON
+	}
+
+	createPolicy = types.MsgCreatePolicy{
+		Creator:      creatorID,
+		Policy:       policy,
+		MarshalType:  policyMarshalType,
+		CreationTime: protoTypes.TimestampNow(),
+	}
+
+	createPolicyResponse, err := l.localModule.GetMsgService().CreatePolicy(
+		l.localModule.GetCtx(),
+		&createPolicy,
+	)
+
+	if err != nil {
+		return "", NewErrFailedToAddPolicyWithACP(err, "Local", creatorID)
+	}
+
+	policyID := createPolicyResponse.Policy.Id
+	log.InfoContext(ctx, "Created Policy", corelog.Any("PolicyID", policyID))
+
+	return policyID, nil
+}
+
+func (l *ACPLocal) ValidateResourceExistsOnValidDPI(
+	ctx context.Context,
+	policyID string,
+	resourceName string,
+) error {
+	if policyID == "" && resourceName == "" {
```

Comment:
This is mostly because I wanted to differentiate between these cases:
1) `@policy`
2) `@policy(id: "id")`
3)  `@policy(resource: "resource")`


For example:
(1) => "missing policy arguments, must have both id and resource"
(2) => "missing policy 'resource' argument"
(3) => "missing policy 'id' argument"

(1) is very easy to mistake so I think it's nice to show all the arguments that are missing.

- Keeping unless there is a hard preference

Reply:
:) I still disagree with you, but will leave this conversation as-is, maybe others will chime in here (not worth a discord ping or @-mention IMO)

Reply:
- [ ] Will revisit, most likely keep as is